synapse/lib/storm.py

```diff
@@ -3165,26 +3174,102 @@ async def execStormCmd(self, runt, genr):
         async with contextlib.AsyncExitStack() as stack:
 
             runts = []
-            for text in self.opts.query:
-                query = await runt.getStormQuery(text)
-                subr = await stack.enter_async_context(runt.getSubRuntime(query))
-                runts.append(subr)
+            query_arguments = await s_stormtypes.toprim(self.opts.query)
+            for arg in query_arguments:
+                if isinstance(arg, str):
+                    arg = (arg, )
+                for text in arg:
+                    query = await runt.getStormQuery(text)
+                    subr = await stack.enter_async_context(runt.getSubRuntime(query))
+                    runts.append(subr)
+            size = len(runts)
+
+            if self.opts.concurrency < 1:
+                raise s_exc.BadArg(mesg='Concurrency must be greater than 0.', concurrency=self.opts.concurrency)
+            semaphore_size = self.opts.concurrency
+            outq_size = semaphore_size * 2
 
             item = None
             async for item in genr:
+                node, path = item
+
+                if self.opts.parallel:
+
+                    outq = asyncio.Queue(maxsize=outq_size)
+                    sempahore = asyncio.BoundedSemaphore(value=semaphore_size)
+                    for subr in runts:
+                        subg = s_common.agen((node, path.fork(node)))
+                        self.runt.snap.schedCoro(self.pipeline(sempahore, outq, subr, genr=subg))
```

Comment:
This will not properly parallelize to a concurrency > queries will it?

Reply:
If concurrency (C) is >= # of queries; they will all run simultaneously.

If concurrency (C) < number of queries, C queries will be able to run simultaneous, and once a query finishes, a pending pipeline async task will be able to obtain the semaphore and start executing, so that up to C queries are running at a given time.